compiler/rustc_passes/src/liveness.rs

```diff
@@ -1686,6 +1710,38 @@ impl<'tcx> Liveness<'_, 'tcx> {
         }
     }
 
+    fn has_added_lit_match_name_span(
+        &self,
+        name: &str,
+        opt_body: Option<&hir::Body<'_>>,
+        err: &mut rustc_errors::DiagnosticBuilder<'_, ()>,
+    ) -> bool {
+        let mut has_litstring = false;
+        if let Some(opt_body) = opt_body {
+            let mut visitor = CollectLitsVisitor { lit_exprs: vec![] };
+            intravisit::walk_body(&mut visitor, opt_body);
+            for lit_expr in visitor.lit_exprs {
+                if let hir::ExprKind::Lit(litx) = &lit_expr.kind {
+                    if let rustc_ast::LitKind::Str(syb, _) = litx.node {
+                        let name_str: &str = syb.as_str();
+                        let mut name_pa = String::from("{");
+                        name_pa.push_str(&name);
+                        name_pa.push('}');
+                        if name_pa.eq(name_str) {
```

Comment:
this only finds those cases where the string is exactly `"{somevar}"`, but to make this more general we'd need to detect `"asdf{somevar:?}bar"`, too. I guess ideally we'd try to parse the string with the regular format string parsing logic and see what variables were captured. Not sure how feasible that is. We could delay it for later and leave the issue open to track improving this further if you'd like.

Reply:
Fixed the case to detect "asdf{somevar:?}bar" too, and updated the test case to detect substring inside. Thanks for the bug report.
I agree that we should parse the string with the regular format string parsing logic and see what variables were captured. But that would make tremendous change for the compiler. i.e, we parse  "{somevar}" as a symbol everywhere, if we want to compare 2 strings, we would need to compare these 2 symbols as well. I would suggest we delay it later. I will open an issue for this one.